crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -7,15 +7,140 @@ use ruma::{
+}
+
+const AUTHENTICITY_NOT_GUARANTEED: &str =
+    "The authenticity of this encrypted message can't be guaranteed on this device.";
+const UNVERIFIED_DEVICE: &str = "Encrypted by an unverified device.";
+const UNSIGNED_DEVICE: &str = "Encrypted by a device not verified by its owner.";
+const UNKNOWN_DEVICE: &str = "Encrypted by an unknown or deleted device.";
+
+impl VerificationState {
+    /// Converts the state to a shield state that can be used by clients
+    pub fn to_shield_state_strict(&self) -> ShieldState {
+        match self {
+            VerificationState::Verified => ShieldState { color: ShieldColor::GREEN, message: None },
+            VerificationState::Unverified(level) => ShieldState {
+                color: ShieldColor::RED,
+                message: match level {
+                    VerificationLevel::UnverifiedIdentity | VerificationLevel::UnsignedDevice => {
+                        Some(UNVERIFIED_DEVICE.to_owned())
+                    }
+                    VerificationLevel::None(link) => match link {
+                        DeviceLinkProblem::MissingDevice => Some(UNKNOWN_DEVICE.to_owned()),
+                        DeviceLinkProblem::InsecureSource => {
+                            Some(AUTHENTICITY_NOT_GUARANTEED.to_owned())
+                        }
+                    },
+                },
+            },
+        }
+    }
+
+    /// Legacy decoration mode.
+    /// Converts the state to a shield state that can be used by clients
+    pub fn to_shield_state_lax(&self) -> ShieldState {
+        match self {
+            VerificationState::Verified => ShieldState { color: ShieldColor::GREEN, message: None },
+            VerificationState::Unverified(level) => match level {
+                VerificationLevel::UnverifiedIdentity => {
+                    // If you didn't show interest in verifying that user we don't
+                    // nag you with an error message
+                    // TODO We should detect identity rotation of a previously trusted identity and
+                    // then warn see https://github.com/matrix-org/matrix-rust-sdk/issues/1129
+                    ShieldState { color: ShieldColor::NONE, message: None }
+                }
+                VerificationLevel::UnsignedDevice => {
+                    // This is a high warning. The sender hasn't verified his own device.
+                    ShieldState {
+                        color: ShieldColor::RED,
+                        message: Some(UNSIGNED_DEVICE.to_owned()),
+                    }
+                }
+                VerificationLevel::None(link) => match link {
+                    DeviceLinkProblem::MissingDevice => {
+                        // Have to warn as it could have been a temporary injected device.
+                        // Notice that the device might just not be known at this time, so callers
+                        // should retry when there is a device change for that user.
```

Comment:
As long as we're mentioning this, we should probably state why it could happen. Presumably federation lag.